<a href="https://colab.research.google.com/github/Hari1331/Hari_DS/blob/master/using_Convolutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Improving Computer Vision Accuracy using Convolutions**

how to do fashion recognition using a Deep Neural Network (DNN) containing three layers 
-- **the input layer **(in the shape of the data)
-- the output layer (in the shape of the desired output)
and a hidden layer.

In [1]:
import tensorflow as tf
print(tf.__version__)

1.14.0


get the Fashion data and load into Train and Test dataset

In [2]:
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()
training_images=training_images / 255.0
test_images=test_images / 255.0
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images, training_labels, epochs=5)

test_loss = model.evaluate(test_images, test_labels)

4423680/4422102 [==============================] - 0s 0us/step


W0731 11:48:38.077512 140325009549184 deprecation.py:506] From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/init_ops.py:1251: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


Epoch 1/5
60000/60000 [==============================] - 4s 74us/sample - loss: 0.4999 - acc: 0.8253
Epoch 2/5
60000/60000 [==============================] - 4s 66us/sample - loss: 0.3783 - acc: 0.8637
Epoch 3/5
60000/60000 [==============================] - 4s 67us/sample - loss: 0.3394 - acc: 0.8760
Epoch 4/5
60000/60000 [==============================] - 4s 65us/sample - loss: 0.3144 - acc: 0.8850
Epoch 5/5
10000/10000 [==============================] - 0s 36us/sample - loss: 0.3352 - acc: 0.8809


So the lost we got it : 0.8913

our accuracy is probably about 89% on training and 87% on validation...not bad...

**Convolution**s

how do you make that even better? One way is to use something called Convolutions
If you've ever done image processing using a filter (like this: https://en.wikipedia.org/wiki/Kernel_(image_processing)) then convolutions will look very familiar.


1.   In short, you take an array (usually 3x3 or 5x5) and pass it over the image. 
2.   By changing the underlying pixels based on the formula within that matrix, you can do things like edge detection.
3.  you'll see a 3x3 that is defined for edge detection where the middle cell is 8,and all of its neighbors are -1.
4.  In this case, for each pixel, you would multiply its value by 8, then subtract the value of each neighbor.
5.  Do this for every pixel, and you'll end up with a new image that has the edges enhanced.

